src/Features/Core/Portable/FileHeaders/AbstractFileHeaderCodeFixProvider.cs

```diff
@@ -0,0 +1,243 @@
+
+        public override FixAllProvider GetFixAllProvider()
+            => new FixAll(this);
+
+        public override Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            foreach (var diagnostic in context.Diagnostics)
+            {
+                context.RegisterCodeFix(
+                    new MyCodeAction(cancellationToken => GetTransformedDocumentAsync(context.Document, cancellationToken)),
+                    diagnostic);
+            }
+
+            return Task.CompletedTask;
+        }
+
+        protected abstract AbstractFileHeaderHelper FileHeaderHelper { get; }
+        protected abstract ISyntaxKinds SyntaxKinds { get; }
+
+        protected abstract SyntaxTriviaList ParseLeadingTrivia(string text);
+
+        private async Task<Document> GetTransformedDocumentAsync(Document document, CancellationToken cancellationToken)
+        {
+            return document.WithSyntaxRoot(await GetTransformedSyntaxRootAsync(document, cancellationToken).ConfigureAwait(false));
+        }
+
+        private async Task<SyntaxNode> GetTransformedSyntaxRootAsync(Document document, CancellationToken cancellationToken)
+        {
+            var tree = await document.GetRequiredSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+            var root = await tree.GetRootAsync(cancellationToken).ConfigureAwait(false);
+
+            var options = document.Project.AnalyzerOptions.AnalyzerConfigOptionsProvider.GetOptions(tree);
+            if (!options.TryGetEditorConfigOption(CodeStyleOptions.FileHeaderTemplate, out string fileHeaderTemplate)
+                || string.IsNullOrEmpty(fileHeaderTemplate))
+            {
+                // Avoid making changes if we fail to read the expected header
+                return root;
+            }
+
+            var expectedFileHeader = fileHeaderTemplate.Replace("{fileName}", Path.GetFileName(document.FilePath));
+
+            var fileHeader = FileHeaderHelper.ParseFileHeader(root);
+            SyntaxNode newSyntaxRoot;
+            if (fileHeader.IsMissing)
+            {
+                newSyntaxRoot = AddHeader(document, root, expectedFileHeader);
+            }
+            else
+            {
+                newSyntaxRoot = ReplaceHeader(document, root, expectedFileHeader);
+            }
+
+            return newSyntaxRoot;
+        }
+
+        private SyntaxNode ReplaceHeader(Document document, SyntaxNode root, string expectedFileHeader)
+        {
+            // Skip single line comments, whitespace, and end of line trivia until a blank line is encountered.
+            var trivia = root.GetLeadingTrivia();
+            var onBlankLine = false;
+            var inCopyright = true;
+            var removalList = new List<int>();
+            var leadingSpaces = string.Empty;
+            var possibleLeadingSpaces = string.Empty;
+
+            // Need to do this with index so we get the line endings correct.
+            for (var i = 0; i < trivia.Count; i++)
```

Comment:
so, can we document this loop.  tbh, as i was reading it i got confused a few times.  i'll add my confusion moments, but some examples/omment could help here.

Reply:
Partially improved. The rest I'm planning to leave for future.